ziti/cmd/ziti/util/certs.go

```diff
@@ -0,0 +1,151 @@
+	"encoding/pem"
+	"fmt"
+	"github.com/fullsailor/pkcs7"
+	"github.com/openziti/ziti/ziti/cmd/ziti/cmd/common"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"reflect"
+)
+
+func WriteCert(p common.Printer, id string, cert []byte) (string, error) {
+	cfgDir, err := ConfigDir()
+	if err != nil {
+		return "", err
+	}
+	certsDir := filepath.Join(cfgDir, "certs")
+	if err = os.MkdirAll(certsDir, 0700); err != nil {
+		return "", errors.Wrapf(err, "unable to create ziti certs dir %v", certsDir)
+	}
+	certFile := filepath.Join(certsDir, id)
+	if err = ioutil.WriteFile(certFile, cert, 0600); err != nil {
+		return "", err
+	}
+	p.Printf("Server certificate chain written to %v\n", certFile)
+	return certFile, nil
+}
+
+func ReadCert(id string) ([]byte, string, error) {
+	cfgDir, err := ConfigDir()
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "couldn't get config dir while reading cert for %v", id)
+	}
+	certFile := filepath.Join(cfgDir, "certs", id)
+	_, err = os.Stat(certFile)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, "", nil
+		}
+		return nil, "", errors.Wrapf(err, "error while statting cert file for %v", id)
+	}
+	result, err := ioutil.ReadFile(certFile)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "error while reading cert for %v", id)
+	}
+	return result, certFile, nil
+}
+
+func IsServerTrusted(host string) (bool, error) {
+	resp, err := http.DefaultClient.Get(fmt.Sprintf("%v/.well-known/est/cacerts", host))
+	if err != nil {
+		if ue, ok := err.(*url.Error); ok && errors.As(ue.Err, &x509.UnknownAuthorityError{}) {
+			return true, nil
+		}
+		return false, err
+	}
+	_ = resp.Body.Close()
+	return false, nil
+}
+
+func GetWellKnownCerts(host string) ([]byte, []*x509.Certificate, error) {
+	transport := *http.DefaultTransport.(*http.Transport)
+	transport.TLSClientConfig = &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	client := http.Client{
+		Transport: &transport,
+	}
+
+	resp, err := client.Get(fmt.Sprintf("%v/.well-known/est/cacerts", host))
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() { _ = resp.Body.Close() }()
+	encoded, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+	certData, err := base64.StdEncoding.DecodeString(string(encoded))
+	if err != nil {
+		return nil, nil, err
+	}
+	certs, err := pkcs7.Parse(certData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	buf := &bytes.Buffer{}
+
+	for _, cert := range certs.Certificates {
+		buf.WriteString("subject=")
+		buf.WriteString(cert.Subject.ToRDNSequence().String())
+		buf.WriteString("\nissuer=")
+		buf.WriteString(cert.Issuer.ToRDNSequence().String())
+		buf.WriteString("\n")
+
+		encoded := pem.EncodeToMemory(&pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: cert.Raw,
+		})
+		buf.Write(encoded)
+	}
+
+	return buf.Bytes(), certs.Certificates, nil
+}
+
+func AreCertsSame(p common.Printer, server, local []byte) bool {
+	serverCerts := decodeCerts(server)
```

Comment:
The order of the certificates in the well-known bundle are not in any order. It is possible to re-order them and have the exact same bundle with the same result.

I don't know if it is worth covering that case, but there can be false positives from the way this is handled now.

Reply:
Ah, ok. Maybe I can order them lexographically to give them a consistent, if arbitrary, order

Reply:
If you order them by their raw bytes, that would cover re-ordering issues for now.

Subject information is only as good as the CA that created it - so I generally avoid using it for anything.

Reply:
Fixed: Added ordering before compare